libraries/automaton/automaton.lua

```diff
@@ -0,0 +1,168 @@
+    end
+
+    return attachment_id
+end
+
+local get_assembly_id = function(part)
+    local assembly_id = nil
+    local assemblies = list(unpack(items:find_ids(part:normalize()) or {}))
+
+    if #assemblies ~= 0 then
+        assembly_id = assemblies:where(function(id) return res_items[id].category == 'Automaton' end)[1] - assembly_offset
+    end
+
+    return assembly_id
+end
+
+local automaton = {
+    remove_all = function()
+        local untraditional_equip = {
+            job = 0x12,
+            is_sub = player.sub_job == 0x12,
+
+            item_index = 0,
+            slots = {}
+        }
+
+        for k, v in pairs(data.attachments) do
+            untraditional_equip.slots[k + 1] = v.id
+        end
+
+        packet.outgoing[0x102][0x12]:inject(untraditional_equip)
+    end,
+
+    equip_head = function(head)
+        if head ~= data.head.name then
+            local head_id = get_assembly_id(head)
+            if head_id and data.available_heads[head_id] then
+                packet.outgoing[0x102][0x12]:inject({
+                    job = 0x12,
+                    is_sub = player.sub_job == 0x12,
+
+                    item_index = head_id,
+                    head = head_id
+                })
+            elseif head_id == nil then
+                print(head .. ' is an invalid head!')
+            else
+                print(head .. ' is not availible!')
+            end
+        end
+    end,
+
+    equip_frame = function(frame)
+        if frame ~= data.frame.name then
+            local frame_id = get_assembly_id(frame)
+            if frame_id and data.available_frames[frame_id - 32] then
+                packet.outgoing[0x102][0x12]:inject({
+                    job = 0x12,
+                    is_sub = player.sub_job == 0x12,
+
+                    item_index = frame_id,
+                    frame = frame_id
+                })
+            elseif frame_id == nil then
+                print(frame .. ' is an invalid frame!')
+            else
+                error(frame .. ' is not availible!')
+            end
+        end
+    end,
+
+    equip_attachment = function(slot, attachment)
+        local attachment_id = get_attachment_id(attachment)
+        if attachment_id and data.available_attachments[attachment_id] then
+            local untraditional_equip = {
+                job = 0x12,
+                is_sub = player.sub_job == 0x12,
+
+                item_index = attachment_id,
+                slots = {0,0,0,0,0,0,0,0,0,0,0,0}
+            }
+
+            untraditional_equip.slots[slot] = attachment_id;
+            packet.outgoing[0x102][0x12]:inject(untraditional_equip)
+        elseif attachment_id == nil then
+            print(attachment .. ' is an invalid attachment!')
+        else
+            print(attachment .. ' is not availible!')
+        end
+    end,
+}
+
+local mt_automaton = {
+    __index = data
+}
+
+return setmetatable(automaton, mt_automaton)
```

Comment:
I thought you were gonna reverse that, so you don't access `data` through this outer table, but instead add the missing functions to the `data` type? Like so:
```lua
ftype.fields.remove_all = {
    data = function(data) -- the data parameter here will be the outer automaton struct
        -- logic
    end,
}
```

I also changed it to take one dummy parameter (`this`) so we can use it similar to our other libraries, i.e. `automaton:remove_all()`. Let me know if you want some explanation on this topic.